Changes

Aeppel opened the pull request.

catjackazz committed:
Build project

catjackazz committed:
Refresh gitignore

catjackazz committed:
Add TestTree

catjackazz committed:
Update Tree code

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Node.java:
```diff
@@ -4,7 +4,8 @@ public class Node {
     protected Node leftNode;
     protected int number;
-    protected int color = 1;
+    protected String color = "Red";
```
Используй enum

Aeppel replied:
done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Node.java:
```diff
@@ -15,15 +16,70 @@ protected Node(int number) {
 
-    protected String show1(Node node) {
-        String result = "";
+    protected StringBuilder showInside(Node node) {
```
showInside немного не говорящий метод и зачем ты постоянно создешь стринг билдеры ? Передавай его в качестве аргумента

Aeppel replied:
Done

Tungrus commented on .idea/gradle.xml:
```diff
@@ -22,6 +22,8 @@
             <option value="$PROJECT_DIR$/base/montecarlo" />
             <option value="$PROJECT_DIR$/base/tickets" />
+            <option value="$PROJECT_DIR$/multithreading" />
+            <option value="$PROJECT_DIR$/multithreading/writersandreaders" />
```
Вот этот файл наверное надо исключить

Aeppel replied:
done

Tungrus replied:
чета нифига не Done) он все еще в проекте )

Aeppel replied:
Done

Tungrus commented on .idea/workspace.xml:
```diff
@@ -4,53 +4,14 @@
+      <change beforePath="$PROJECT_DIR$/algorithms/tree/src/main/java/org/pervishkond/tree/Node.java" beforeDir="false" afterPath="$PROJECT_DIR$/algorithms/tree/src/main/java/org/pervishkond/tree/Node.java" afterDir="false" />
+      <change beforePath="$PROJECT_DIR$/algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java" beforeDir="false" afterPath="$PROJECT_DIR$/algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java" afterDir="false" />
+      <change beforePath="$PROJECT_DIR$/algorithms/tree/src/test/java/org/pervishkond/tree/TestTree.java" beforeDir="false" afterPath="$PROJECT_DIR$/algorithms/tree/src/test/java/org/pervishkond/tree/TestTree.java" afterDir="false" />
+      <change beforePath="$PROJECT_DIR$/settings.gradle" beforeDir="false" afterPath="$PROJECT_DIR$/settings.gradle" afterDir="false" />
```
Вот эти дифы наверное тоже не надо

Aeppel replied:
done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Main.java:
```diff
@@ -4,76 +4,6 @@ public class Main {
     public static void main(String[] args) {
 
         Tree tree = new Tree();
```
Какой то огрызок остался, может вообще его уберешь если он не нужен ?

Aeppel replied:
done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Node.java:
```diff
@@ -15,15 +16,70 @@ protected Node(int number) {
+        nodeRoot.setColor("Black");
+    }
+
+    protected Node whoIsParent(Node node, Node nodeRoot) {
```
1 - очень странное название, 
2 ты должен работать с this и 1 аргументом. Очень странно такой код писать внутри класса если он не использует его внутренние свойства

Aeppel replied:
done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -7,97 +7,62 @@ class Tree extends Node {
+        if (empty()) {
             nodeRoot = new Node(number);
-            nodeRoot.color = 0;
+            nodeRoot.setColor("Black");
```
ну ты понял,enum

Aeppel replied:
done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -617,10 +562,109 @@ private Node checkingNephew(Node node) {
+        Node checkingBadNode;
+        Node nodePointer = nodeRoot;
+        Node addNode = new Node(number);
+        while (true) {
```
рекурсивный поиск надо делать. циклы в дереве дичь

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -617,10 +562,109 @@ private Node checkingNephew(Node node) {
+        node.leftNode.setColor("Red");
+    }
 
+    private void putIntoTree(int number) {
```
чел тут просто пиздец ) Давай полегче. Постарайся.

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -617,10 +562,109 @@ private Node checkingNephew(Node node) {
+        Node nodePointer = nodeRoot;
+        Node addNode = new Node(number);
+        while (true) {
+            if (nodePointer.number > addNode.number && nodePointer.leftNode != null) {
```
в метод

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -617,10 +562,109 @@ private Node checkingNephew(Node node) {
+        while (true) {
+            if (nodePointer.number > addNode.number && nodePointer.leftNode != null) {
+                nodePointer = nodePointer.leftNode;
+            } else if (nodePointer.number < addNode.number && nodePointer.rightNode != null) {
```
в метод

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -7,97 +7,62 @@ class Tree extends Node {
 
             Node checkingBadNode = checkLeft(nodeRoot);
-            if (checkingBadNode != null) {
+            if (isNotNull(checkingBadNode)) {
```
bad node это не какой то понятный термин вообще не ясно что значит bad в твоем понимании. давай подумай еще

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -389,101 +321,110 @@ private Node searchOfMaxNumber(Node node) {
+                        if (isBlack(nodePointer.leftNode) && isBlack(nodePointer.rightNode)) { // TODO: 06.02.2023  вывести эту хрень в отдельный метод, параметром сделать nodePointer также сделать это и для другой стороны
+                            Node nodePointerChecker = nodePointer.leftNode;
+                            if (isNotNull(nodePointerChecker.leftNode) && isRed(nodePointerChecker.leftNode))
+                                nodePointer.setColor("Red");
```
enum

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/test/java/org/pervishkond/tree/TestTree.java:
```diff
@@ -0,0 +1,120 @@
+
+    @BeforeEach
+    public void setUp() {
+        attempt1 = 50;
```
бро ну зачем ? почему не просто константа ?)

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/test/java/org/pervishkond/tree/TestTree.java:
```diff
@@ -0,0 +1,120 @@
+        testTree.add(attempt5);
+        testAdd();
+        testRemove(attempt3);
+        testRemove(attempt3);
```
test это преписка имен методов помеченных аннотацией @Test. 
их нельзя вызывать в других методах.

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/test/java/org/pervishkond/tree/TestTree.java:
```diff
@@ -0,0 +1,120 @@
+    }
+
+
+    private void checkOfAddingNodeRoot() {
```
методы которые делает assert называются assert

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -7,97 +7,62 @@ class Tree extends Node {
         }
-        if (parent.rightNode != null) {
-            parent.rightNode.color = 0;
+        if (checkForChild(node, parent)) {
```
check не очень понятная преписка методам, такие методы делаются isBigCock isTeamNoob

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -108,20 +73,19 @@ protected void rotateLeft(Node node, Node parent) {
                 nodePointer.rightNode = node.leftNode;
                 nodeRoot.leftNode = nodePointer;
-                nodeRoot.rightNode.color = 0;
+                nodeRoot.rightNode.setColor("Black");
```
getRightNode() не надо обращаться к полям напрямую

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -495,114 +436,118 @@ private void balancingAfterRemoving(Node node) { //parent
         }
     }
 
     private void reverseRecoloring(Node parent) {
```
в class Node

Aeppel replied:
done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Node.java:
```diff
@@ -15,15 +16,70 @@ protected Node(int number) {
+        this.color = color;
+    }
+
+    protected void recoloring(Node parent, Node nodeRoot) {
```
в class Node

Aeppel replied:
done

catjackazz committed:
Update Adding Function, Use enum for colors

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Colors.java:
```diff
@@ -0,0 +1,5 @@
+package org.pervishkond.tree;
+
+public enum Colors {
+    RED, BLACK
```
Обычно их делают 1 объявление на 1 строчке. 
RED,
BLACK

Aeppel replied:
Done

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Main.java:
```diff
@@ -2,78 +2,6 @@
 
 public class Main {
```
Пустой мейн наверное можно удалить целиком

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Node.java:
```diff
@@ -4,8 +4,10 @@ public class Node {
     protected Node leftNode;
     protected int number;
-    protected int color = 1;
+    //    protected String color = "Red";
```
такое нельзя допускать, выкидывай строчку целиком перед коммитом

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -2,274 +2,208 @@
             if (parent == nodeRoot) {
                 nodeRoot = node;
-                nodePointer.rightNode = node.leftNode;
+                nodePointer.rightNode = node.getLeftNode();
```
мы вроде решили обращатсья к полям по через методы ) 
а то у тебя тут играй тут не играй тут рыбу заворачивали

Aeppel replied:
Done

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -2,274 +2,208 @@
 
-    public String show() {
-        return show1(nodeRoot);
+    public void show() {
```
А в чем смысол ?) Процессору не дать заскучать ?)
Подсказка есть прекрасный метод toString у Object и возможно ты захочешь реализовать его для решения задачи отрисовки дерева )

Aeppel replied:
Done

Tungrus replied:
Так, искренне не понял что тут Done.

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Node.java:
```diff
@@ -15,15 +17,182 @@ protected Node(int number) {
-        String result = "";
-        if (leftNode != null) {
-            result = leftNode.show1(node.leftNode);
+    protected StringBuilder showTree(StringBuilder result) {
```
Я вот тут не знаю как правильно изобразить ситуэшен, но ты можешь использовать перепорпеделение уже существующих методов toString от Object и добавить реализацию вида StringBuilder toString(StringBuilder str).
Вызов String toString() дергает рекурсивный метод StringBuilder toString(StringBuilder)
Это ближе к вкусовщине но мне кажется тогда не образуется методов аппендиксов которые ничего не делают.

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Node.java:
```diff
@@ -4,8 +4,10 @@ public class Node {
-    protected int color = 1;
+    //    protected String color = "Red";
 
+    protected Colors color = Colors.RED;
```
Все параметры вроде null в начале а этот ты почему то выставляешь изначально, почему не написать пару конструкторов ?

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Node.java:
```diff
@@ -15,15 +17,182 @@ protected Node(int number) {
-        result += " " + this.number;
-        if (rightNode != null) {
-            result += rightNode.show1(node.rightNode);
+        result.append(" ").append(this.number);
```
В java есть мега полезный вариант StringBuidler - StringJoiner ты ему указываешь например складывать через запятые(в констуркторе) и дальше просто делаешь ему просто add. 
![image](https://user-images.githubusercontent.com/17014717/235363597-7a728659-fdcb-4b72-be55-91addfd4471a.png)

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Node.java:
```diff
@@ -15,15 +17,182 @@ protected Node(int number) {
+            this.getRightNode().setColor(Colors.BLACK);
+        }
+        nodeRoot.setColor(Colors.BLACK);
+    }
```
как то странно метод выглядит, почему он назначает цвет себе, или какой то другой непонятной ноде которая к нему никак не отностися, надо поправить. Если он захочет рекурсией перекрашивать все дерево пожалуйста. Но вот это как то все странно.

Aeppel replied:
Потому что он меняет себе цвет и своим детям

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Node.java:
```diff
@@ -15,15 +17,182 @@ protected Node(int number) {
+        nodeRoot.setColor(Colors.BLACK);
+    }
+
+    protected void reverseRecoloring() {
```
Вот этот метод выглядит вполне валидным

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Node.java:
```diff
@@ -15,15 +17,182 @@ protected Node(int number) {
+        return node != null;
+    }
+
+    // Блок условий optionToBalance
```
Вот такие штуки поровней делать их над методом с которого начинаешь и // --------- Блок условий optionToBalance ---------. Остальные места не помечаю найди и поправь.

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Node.java:
```diff
@@ -15,15 +17,182 @@ protected Node(int number) {
+        return node.color == Colors.BLACK;
+    }
+
+    protected Boolean hasChild(Node node, Node parent) {
```
Внимательней смотрим такие места ) можно объединить условия через ||

Aeppel replied:
Done (Метод удален в связи с ненадобностью)

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Node.java:
```diff
@@ -15,15 +17,182 @@ protected Node(int number) {
+    }
+
+    protected void recoloringAfterRemoving() {
+        Node parent = this.searchParentNode(this.nodeRoot);
```
Ты завел nodeRoot который по смыслу parent для текущей node. скажи зачем ты обращаешься к методу ?

Aeppel replied:
Done( Добавил в параметры метода nodeRoot - дерева, потому что метод берет данную ноду из поля node, которое null  изначально)

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Node.java:
```diff
@@ -15,15 +17,182 @@ protected Node(int number) {
+    }
+
+    protected void recoloringAfterRemoving() {
+        Node parent = this.searchParentNode(this.nodeRoot);
```
this то используешь, то не используешь, наводишь путаницу. Старайся прямое упоминание this не использовать.

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Node.java:
```diff
@@ -15,15 +17,182 @@ protected Node(int number) {
+            this.getLeftNode().setColor(Colors.BLACK);
+        } else {
+            if (this.getLeftNode() != null) this.getLeftNode().setColor(Colors.RED);
+            else this.getRightNode().setColor(Colors.RED);
```
Вкусвщина но выглядит без фигурных скобок очень неаккуратно

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Node.java:
```diff
@@ -15,15 +17,182 @@ protected Node(int number) {
+        }
+    }
+
+    protected void recoloringAfterRemoving() {
```
Надо подумать над упрощением этой партянки, читать сложно

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -2,274 +2,208 @@
 
     }
 
     private void optionsToBalance(Node node) {
```
Не выкупаю как понимать название метода ) Типа условия балансирования дерева ? Так просто назови его rebalance

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -2,274 +2,208 @@
-        if (parent.leftNode != null) {
-            parent.leftNode.color = 0;
+    protected void rotateLeft(Node node, Node parent) {
+        if (node == null || parent == null) {
```
Какое то странное условие ) Типа ты хочешь повернуть дерево и даешь 2 аргумента и проверяешь их в методе. Обычно стараются зоны ответственности разграничивать и если метод поворачивает то он только поворачивает. По хорошему ты не должен в такие ситуации поппасть.

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -2,274 +2,208 @@
-        if (node.leftNode == null || node.leftNode.color == 0) {
-            node.color = 0;
-            parent.color = 1;
+        Node ln = node.leftNode;
```
Убери переменную тем более с таким именем )

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -2,274 +2,208 @@
-
-
-        } else if (parent.leftNode == null) {
+        } else if (isNull(parent.getLeftNode()) || parent.getLeftNode().number == 0) {
```
Ошибка ?)

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -2,274 +2,208 @@
-        } else if (node.rightNode == null || node.rightNode.color == 0) { /// TODO: 21.11.2022 разобраться с малым поворотом на левую сторону
-            Node nodePointer = node.leftNode.rightNode;
+        } else if (isRightNodeBlack(node)) {
+            Node nodePointer = node.getLeftNodeRightChild();
```
Зачем ты используешь переменные которые тебе не нужны ?)

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -2,274 +2,208 @@
+            parent.setColor(Colors.RED);
             node.rightNode = parent;
-            parent.leftNode = n;
+            parent.leftNode = rn;
```
Переменная не используется почти надо убрать

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Node.java:
```diff
@@ -15,15 +17,182 @@ protected Node(int number) {
+    }
+
+    //Блок условий CheckLeft/Right
+    protected boolean isGoingLeft(Node node) {
```
Не понял названия, давай подумаем как назвать, надо качать качественный нейминг переменных

Aeppel replied:
Done

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -366,263 +300,275 @@ public void remove(int number) {
-
     }
 
     private void swap(Node maxNode, Node removeNode, Node parent) {
```
Максимальная Нода )) неее давай подумаем как это еще назвать )

Aeppel replied:
Оставил изначальную, ссылаясь на статью

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -366,263 +300,275 @@ public void remove(int number) {
-                reverseRecoloring(whoIsParent(parent, nodeRoot));
+            if (isRed(parent.leftNode.rightNode) && isRed(parent.leftNode.leftNode)) {
+                node.leftNode.setColor(Colors.BLACK);
+            } else if (isRed(parent.getLeftNodeRightChild()) || isRed(parent.getLeftNodeLeftChild())) {
```
Враиации LeftLeft RightRight и прочее не надо использовать, в 1 случае еще ок. Но тут тебе надо со стороны бизнес логики говорить (если так говорить допустимо в нашем случае) что происходит В красночерном дереве есть термины, надо от них отталкиваться

Aeppel replied:
Done ( не избежать данного обозначения, т.к. тех ж внуков 4 вариации, но добавил GrandChild  и Child)

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -366,263 +300,275 @@ public void remove(int number) {
-                parent.rightNode.leftNode = null;
-            if (parent.rightNode.rightNode != null && parent.rightNode.rightNode.number == 0)
-                parent.rightNode.rightNode = null;
+            if (isNotNull(child.getLeftNode()) && child.getLeftNode().number == 0)
```
да что означает это условие ?)

Aeppel replied:
Обусждали- создание  нулевой ноды внутри метода

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -366,263 +300,275 @@ public void remove(int number) {
-                                nodePointer.color = 1;
-                            if (nodePointer.rightNode.rightNode != null && nodePointer.rightNode.rightNode.color == 1)
-                                nodePointer.color = 1;
+                if ((isNotNull(node.getRightNode()) && isRed(node.getRightNode())) || (isNotNull(node.getLeftNode()) && isRed(node.getLeftNode())))
```
уффф надо подумать над упрощением вот такого трындеца даже читать боюсь )

Aeppel replied:
Done

Aeppel replied:
Done, вывел в отдельный метод

catjackazz committed:
Update Code

catjackazz committed:
Update code

catjackazz committed:
Update Code

catjackazz committed:
Update Code

Tungrus commented on algorithms/tree/src/test/java/org/pervishkond/tree/TestTree.java:
```diff
@@ -0,0 +1,109 @@
+
+    @Test
+    public void setTestTree() {
+        testTree.add(attempt1);
```
Надо будет созвониться по поводу этого кода

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -2,360 +2,284 @@
-            nodeRoot.color = 0;
+    public void add(int number) {
+        if (isEmpty()) {
+            setNodeRoot(new Node(number));
```
Можно вынести новую ноду в переменную и сделать все дела.
Можно расширить конструктор и не делать лишний сет.

Aeppel replied:
done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -366,263 +290,276 @@ public void remove(int number) {
 
+    private boolean isEmpty() {
+        return size == 0;
     }
```
Стандартный метод для коллекций так то, public

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -366,263 +290,276 @@ public void remove(int number) {
+        } else if (isEqual(addNode, checkNode)) {
+            size--;
+            return null;
+        }
```
Это как то странно выглядит. У тебя же не Set(Множество). По хорошему Node должен хранить количество своих копий.
Может быть я тут не прав. 
Если оставлять этот варинт то тоже плохо. У тебя есть return работай с ним в методе add не надо size туда сюда дергать

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -366,263 +290,276 @@ public void remove(int number) {
+            size--;
+            return null;
+        }
+        if (addNode == null) {
```
Мегаплохо. У тебя не должно быть ситуаций когда ты ищешь куда тебе пихнуть null

Tungrus replied:
АА, я понял, с толку сбил тем что использовал 1 переменную. Не надо переменную с addNode затирать. 
В разные переменные запиши

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -366,263 +290,276 @@ public void remove(int number) {
+        if (addNode == null) {
+            return null;
+        }
+        if (checkNode.number > addNode.number) {
```
Почему метод поиска, занимается чем то помимо поиска ?

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -366,263 +290,276 @@ public void remove(int number) {
+        } else {
+            checkNode.setRightNode(addNode);
+        }
+        Node NodeForBalance;
```
Кривое имя

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -366,263 +290,276 @@ public void remove(int number) {
+        nodeRoot = node;
+    }
+
+    private Node searchPlaceToPut(Node addNode, Node checkNode) {
```
Надо будет созвониться по поводу этого метода в целом.2

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -2,360 +2,284 @@
-    private void optionsToBalance(Node node) {
-        if (node == nodeRoot) {
-            recoloring(nodeRoot);
+    private void rebalance(Node node) {
```
Старайся сделать свой код максимально читемым.
Нужно придерживаться определенного порядка методов. Сначала публичные потом protected потом private.

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -2,360 +2,284 @@
+    public void add(int number) {
+        if (isEmpty()) {
+            setNodeRoot(new Node(number));
+            getNodeRoot().setColor(Colors.BLACK);
```
В этом месте ты можешь обратиться к полю напрямую (оно приватное но доступ до него у тебя есть

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -366,263 +290,276 @@ public void remove(int number) {
+        if (isNull(nodePointer)) {
             return null;
         }
         if (number == node.number) {
```
Можно объединить через ||

Tungrus replied:
if (isNull(nodePointer) || number == node.number) {
            return nodePointer;

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -2,360 +2,284 @@
-        if ((removeNode.color == 1) && (removeNode.leftNode == null && removeNode.rightNode == null)) {
-            if (parent.leftNode == removeNode) {
-                parent.leftNode = null;
+        Node parent = removeNode.searchParentNode(getNodeRoot());
```
То есть ты нашел удаляемый элемент, а после пошел искать родителя ( ну то есть по 2 кругу поиск запустил). Ищи сразу а то производительность на все этом смотри как то так
![image](https://github.com/Aeppel/Education/assets/17014717/4a70b399-9f0f-4751-b7a8-eb8cec0ee12b)

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -2,360 +2,284 @@
-            if (maxNode == removeNode.leftNode) {
+        if (removeNode.getLeftNode() != null && removeNode.getRightNode() != null) {
+            Node maxNode = searchOfMaxNumber(removeNode.getLeftNode());
+            if (maxNode == removeNode.getLeftNode()) {
```
Тут ты проверяешь что у removeNode.getLeftNode().getLeftNode() == null по сути.

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -2,360 +2,284 @@
                 return;
             }
-            Node parentMaxNode = whoIsParent(maxNode, nodeRoot);
+            Node parentMaxNode = maxNode.searchParentNode(getNodeRoot());
```
у тебя есть nodeRoot у Node зачем ты ищешь по всему дереву ?)

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -2,360 +2,284 @@
+        getNodeRoot().showTree(result);
     }
 
     public void remove(int number) {
```
В целом весь этот метод описывает 4 случая удаления элемента из дерева. Котоыре можно спокойно побить на 4 метода с аргументами которые метод remove будет вызывать. 
Глазами ты можешь посомтреть что утебя 4 независмых if условия (не else if )

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -2,360 +2,284 @@
         }
         size++;
-        return true;
 
```
Следим за пробельчиками, скачут по всему коду то тут там

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -2,360 +2,284 @@
-            recoloring(nodeRoot);
+    private void rebalance(Node node) {
+        if (node == getNodeRoot()) {
+            getNodeRoot().recoloring(getNodeRoot());
```
Надо будет обсудить, зачем этому методу этот аргумент ?

Aeppel replied:
В него падает конфликтная нода потом выбирается метод, балансирующий дерево (перекрас или поворот)

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -2,360 +2,284 @@
 
 class Tree extends Node {
```
Не надо наследоваться

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Node.java:
```diff
@@ -4,26 +4,221 @@ public class Node {
+
+    protected boolean isNull(Node node) {
+        return node == null;
+    }
```
Добавь static и вынеси в отдельный класс с public доступом

Aeppel replied:
Done

catjackazz committed:
Add Utils class, replace NodeRoot to Tree class, add parent in Node class field

catjackazz committed:
Change Search and Put elements into Tree methods

catjackazz committed:
Replace put method into Utils, update code

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Colors.java:
```diff
@@ -0,0 +1,3 @@
+package org.pervishkond.tree;
+
+public enum Colors {RED, BLACK}
```
Есть правила оформления для enum. 
RED и BLACK должны быть на разных строчках. Можешь загуглить если не очень понятен комментарий или спроси

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Node.java:
```diff
@@ -1,29 +1,150 @@
+import java.util.StringJoiner;
+
 public class Node {
     protected Node rightNode;
```
Нужно стремиться уменьшать доступ до переменных. Если есть возможность сделать ее приватной - делай

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Node.java:
```diff
@@ -1,29 +1,150 @@
+        setColor(Colors.RED);
+        this.number = number;
+    }
+
```
В Java есть возможность вызвать другой (не super) конструктор через выражение this(arg1, arg2). Не стоит их полностью вот так реализовывать

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Node.java:
```diff
@@ -1,29 +1,150 @@
+
+    protected void recoloring(Node nodeRoot) {
+        this.setColor(Colors.RED);
+        if (this.getLeftNode() != null) {
```
Я вот не помню тут детали алгоритма, но почему не дернуть recoloring тут для детей ?

Aeppel replied:
У меня всего 3 варианта recoloring (сам recoloring, обратный recoloringб и после удаления)

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Node.java:
```diff
@@ -1,29 +1,150 @@
+            this.getLeftNode().setColor(Colors.BLACK);
+        }
+        if (Utils.isNotNull(this.getRightNode())) {
+            this.getRightNode().setColor(Colors.BLACK);
```
В таких местах не обязательно писать this. Хотя мне в целом нравится как это выглядит. На твое усмотрение что с этим делать

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Node.java:
```diff
@@ -1,29 +1,150 @@
+    }
+
+    protected void deleteRecoloring(Node nodeRoot) {
+        Node parent = this.getParent();
```
Обращай внимание на замечания статического анализатора, ты перекрываешь уже существующую тут переменную parent объявленную на 11 строке.

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Node.java:
```diff
@@ -1,29 +1,150 @@
+        }
+    }
+
+    protected void deleteRecoloring(Node nodeRoot) {
```
Эта переменная тут не нужна

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Node.java:
```diff
@@ -1,29 +1,150 @@
+        if (Utils.isNotNull(grandparent)) {
+            if (grandparent.getLeftNode() == parent) {
+                parent.color = grandparent.getRightNode().color;
+            } else parent.color = grandparent.getLeftNode().color;
```
Когда у тебя идет такая плотная "Застройка "ифами не экономь фигурные скобки, очень легко запутаться.
Еще 1 пустой строчкой можно отделять логическе блоки и if друг от друга что бы немного упростить чтение.

Aeppel replied:
Done

Tungrus replied:
Давай внимательней

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Node.java:
```diff
@@ -1,29 +1,150 @@
+        if (parent.getLeftNode() == this) {
+            this.color = parent.getRightNode().color;
+        } else this.color = parent.getLeftNode().color;
+        if (Utils.isNotNull(this.getLeftNode()) && Utils.isNotNull(this.getRightNode())) {
```
Заведи метод isLeaf(Node node); Leaf - листок это на краях дерева обозначается. Node это у кого есть хотя бы 1 ребенок

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Node.java:
```diff
@@ -1,29 +1,150 @@
+        return color;
+    }
+
+    public void setLeftNode(Node node) {
```
Обычно set get операции группируют либо по признаку set get. Либо по принадлежности к полю и тогда они чередуются

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -1,628 +1,569 @@
-            if (checkingBadNode != null) {
-                optionsToBalance(checkingBadNode);
+            Node parentNode = searchPlace(new Node(number), getNodeRoot());
+            if (Utils.isNotNull(parentNode)) {
```
Можно уменьшить вложенность сделав 
if ( Utils.isNull(parentNode)) {
return;
}

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -1,628 +1,569 @@
-            checkingBadNode = checkLeft(nodeRoot);
-            if (checkingBadNode != null) {
-                optionsToBalance(checkingBadNode);
+            Node parentNode = searchPlace(new Node(number), getNodeRoot());
```
searchPlaceToPut иначе не очень понятно что за место ты там ищешь

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -1,628 +1,569 @@
     }
 
 
     private Node search(int number, Node node) {
```
Можно обойтись без использования этого аргумента для запуска поиска.
Заведи 2 метода search с разными сигнатурами

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -1,628 +1,569 @@
     public void remove(int number) {
-        Node removeNode = search(number, nodeRoot);
+        Node removeNode = search(number, getNodeRoot());
         if (removeNode == null) {
```
Utils.isNull(removeNode)

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -1,628 +1,569 @@
-            if (parent.leftNode == removeNode) {
-                parent.leftNode = null;
+        Node parent = removeNode.getParent();
+        if ((Utils.isRed(removeNode)) && (removeNode.getLeftNode() == null && removeNode.getRightNode() == null)) {
```
аналогично но спецификой проверки

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -1,628 +1,569 @@
+
+
+    private void checkForRebalance() {
+        Node NodeForBalance = checkRight(nodeRoot);
```
Поправь имя

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Node.java:
```diff
@@ -1,29 +1,150 @@
-            result = leftNode.show1(node.leftNode);
+    protected StringJoiner showTree(StringJoiner result) {
+        if (this.getLeftNode() != null) {
+            result.merge(this.getLeftNode().showTree(result));
```
Очень подозрительно что ты подаешь этот метод во внутрь(работаешь с ним через условый add) и на выходе тоже его добавляешь ? 
он случай не нарисует лишнего ?
Накинь тест на этот метод (запиливаешь дерево и сравниваешь его со строчкой на выходое)

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Node.java:
```diff
@@ -1,29 +1,150 @@
-        if (rightNode != null) {
-            result += rightNode.show1(node.rightNode);
+        result.add(String.valueOf(this.number));
+        if (this.getRightNode() != null) {
```
Мне кажется лучше будет если Node сам сможет тебе говорить есть у него левый или правый элемент

Aeppel replied:
Done

catjackazz committed:
Update code

Исправлен уровень доступа метода IsEmpty()
Изменил места методов по нисходящей (public, protected, private)
Переместил метод isNull  в Utils
Исправил строки в enum
Исправил работу метода show
Удалены и исправлены переменные parent исправлены на метод getParent()
Удалена переменная из метода deleteRecoloring
Введен новый метод isChildren для проверки листков (Вместо isLeaf)
Сгруппированы методы set/get

catjackazz committed:
Update code

Разбит на части метод remove()
Разбит на части метод balancingAfterRemoving()

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Node.java:
```diff
@@ -1,29 +1,153 @@
     }
 
-    protected Node(int number) {
+    public Node(int number) {
```
1. Конструкторы обычно объявляют после полей.
2. Зачем тебе protected конструктор ? 
3. Ты не понял как сделать через this. Давай еще раз или спроси или исправь

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Node.java:
```diff
@@ -1,29 +1,153 @@
+
+    protected void recoloring(Node nodeRoot) {
+        this.setColor(Colors.RED);
+        if (this.getLeftNode() != null) {
```
Давай уже к 1 стилю приведем ) через Utils.isNotNull

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Node.java:
```diff
@@ -1,29 +1,153 @@
+        }
+    }
+
+    protected void deleteRecoloring() {
```
Зачем удалять перекрашивание ?) Нужно немного другое имя выбрать

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Node.java:
```diff
@@ -1,29 +1,153 @@
+            this.getRightNode().setColor(Colors.BLACK);
+            this.getLeftNode().setColor(Colors.BLACK);
+        } else {
+            if (this.getLeftNode() != null) {
```
Пройдись по всем файлам и поправь на 1 стиль

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Node.java:
```diff
@@ -1,29 +1,153 @@
+        if (Utils.isNotNull(grandparent)) {
+            if (grandparent.getLeftNode() == parent) {
+                parent.color = grandparent.getRightNode().color;
+            } else parent.color = grandparent.getLeftNode().color;
```
Вот этот иф, иф иф с 127 по 131 и иф с 145 по 149 так то одна и та же операция. Запиши это все в 1 функцию с подачей аргументов

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -1,628 +1,590 @@
-            checkingBadNode = checkLeft(nodeRoot);
-            if (checkingBadNode != null) {
-                optionsToBalance(checkingBadNode);
+            Node parentNode = searchPlaceToPut(new Node(number), getNodeRoot());
```
Вот тут я бы либо джавадок написал к методу searchPlaceToPut или написал метод приватный что у нас уже содержится элемент. Если посмотреть реализацию add то она как правило возвращает boolean об успешности/не успешности добавления эелемента в коллекцию

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -1,628 +1,590 @@
-    }
+class Tree {
+
+    Node nodeRoot;
```
Если я пишу замечание в 1 месте, я ожидаю что ты это правило будешь смотреть по всему коду самостоятельно. private /

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -1,628 +1,590 @@
-    private void optionsToBalance(Node node) {
-        if (node == nodeRoot) {
-            recoloring(nodeRoot);
+    private void rebalance(Node node) {
```
Про сортировку методов не забывай )

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -1,628 +1,590 @@
-        if (parent.rightNode != null) {
-            parent.rightNode.color = 0;
+        size--;
+        if ((Utils.isRed(removeNode)) && (removeNode.getLeftNode() == null && removeNode.getRightNode() == null)) {
```
У тебя в Utils был метод isChildren

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -1,628 +1,590 @@
+        size--;
+        if ((Utils.isRed(removeNode)) && (removeNode.getLeftNode() == null && removeNode.getRightNode() == null)) {
+            removeRedNodeWithNoChildren(removeNode);
+            return;
```
Не получается 4 else if написать ?

Aeppel replied:
Done

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -1,628 +1,590 @@
-            node.leftNode = parent;
-            parent.rightNode = n;
-            if (parent == nodeRoot) {
+    private void rotateLeft(Node node, Node parent) {
```
я бы переписал этот метод в класс Node что бы метод мог возвращать Node. Тогда если он не Null то ты меняешь rootNode на тот который указываешь в качестве возвращаемого значения
Аналогично для правого поворота

Tungrus commented on algorithms/tree/src/main/java/org/pervishkond/tree/Tree.java:
```diff
@@ -1,628 +1,590 @@
-            badNode = checkRight(node.leftNode);
-        } else if ((node.leftNode != null) && (node.leftNode.color == 0)) {
-            badNode = checkRight(node.leftNode);
+    private Node checkRight(Node node) {
```
Думаю это тоже принадлежит классу Node

Aeppel replied:
Done

catjackazz committed:
Update code

Перенесы методы rotateLeft() и rotateRight()  в класс Node
Добавлена обработка nodeRoot в метод recoloring()

catjackazz committed:
Final version

catjackazz committed:
Final version

catjackazz committed:
Final version

catjackazz committed:
Final version
